### ISSUE TITLE
No associated item found when the associated item was used as const generic parameter in return type

### ISSUE THREAD

**Issue by Neo-Zhixing** [rust-lang/rust#99705]
I tried this code:

```rust
// build-pass
#![crate_type = "lib"]
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]
trait MyIterator {
    type Output;
}

trait Foo {
    const ABC: usize;
}

struct IteratorStruct<const N: usize>{

}

struct BarStruct<const N: usize> {
    data: [usize; N]
}

impl<const N: usize> MyIterator for IteratorStruct<N> {
    type Output = BarStruct<N>;
}

fn test<T: Foo>() -> impl MyIterator<Output = BarStruct<{T::ABC}>> where [(); {T::ABC}]: Sized {
    IteratorStruct::<{T::ABC}>{}
}

```

This code should compile normally.

Instead, this happened:

```
error[E0599]: no associated item named `ABC` found for type parameter `T` in the current scope
  --> /home/neo/rust/src/test/ui/const-generics/generic_const_exprs/my_test.rs:28:61
   |
LL | fn test<T: Foo>() -> impl MyIterator<Output = BarStruct<{T::ABC}>> {
   |         -                                                   ^^^ associated item not found in `T`
   |         |
   |         associated item `ABC` not found for this type parameter
   |
   = help: items from traits can only be used if the type parameter is bounded by the trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
```

In the `test` function, T was already bounded by the `Foo` trait which does have an associated item named `ABC`. However, Rust complains that no associated item named `ABC` was found for `T`.

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0-nightly (62b272d25 2022-07-21)
binary: rustc
commit-hash: 62b272d25c5bb8b6bb8ac73797d82b8b9a1eabda
commit-date: 2022-07-21
host: x86_64-pc-windows-msvc
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by Neo-Zhixing** [comment]
I did some digging and this might be related to https://github.com/rust-lang/rust/blob/e33cc71a61c91e1d510bf283e9d345067e64eed2/compiler/rustc_typeck/src/collect.rs#L2468-L2480

@BoxyUwU Mind taking a look?

**Comment by Neo-Zhixing** [comment]
#76560

A fix exists here at #99801 but we're waiting on @spastorino 's pending RPIT refactor work to fix this. In the mean time, developers who need this feature can use the branch in #99801.

https://rust-lang.github.io/rustup/concepts/toolchains.html#custom-toolchains
